### PR TITLE
[ginkgo] Update to 1.6.0

### PR DESCRIPTION
**Pull request by upsj**
Based on #27792, this adds the latest version

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] SHA512s are updated for each updated download
- [x] The "supports" clause reflects platforms that may be fixed by this new version
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by upsj** [comment]
x64_uwp fails because of SDL turning deprecation warnings into errors https://stackoverflow.com/questions/55007377/deprecated-results-in-build-failure-in-vs2017-error-c4996

The other issues I cannot reproduce locally, so no idea what is going wrong there.

File: ports/ginkgo/portfile.cmake [hunk 1]
```diff
@@ -1,27 +1,27 @@
...
 
+set(ADDITIONAL_FLAGS)
+
+if(VCPKG_TARGET_IS_WINDOWS)
+    set(ADDITIONAL_FLAGS "-DCMAKE_CXX_FLAGS_DEBUG='/MDd /Zi /Ob1 /O1 /Od /RTC1'")
+endif()
+
 vcpkg_cmake_configure(
     SOURCE_PATH "${SOURCE_PATH}"
+    WINDOWS_USE_MSBUILD
```
Comment (BillyONeal) [inline ports/ginkgo/portfile.cmake:24]: This is a significant usability regression. Can you describe why the msbuild backend is necessary here?

File: ports/ginkgo/portfile.cmake [hunk 2]
```diff
@@ -1,27 +1,27 @@
...
     FEATURES
     openmp    GINKGO_BUILD_OMP
     cuda      GINKGO_BUILD_CUDA
+    mpi       GINKGO_BUILD_MPI
 )
 
+set(ADDITIONAL_FLAGS)
+
+if(VCPKG_TARGET_IS_WINDOWS)
+    set(ADDITIONAL_FLAGS "-DCMAKE_CXX_FLAGS_DEBUG='/MDd /Zi /Ob1 /O1 /Od /RTC1'")
```
Comment (BillyONeal) [inline ports/ginkgo/portfile.cmake:19]: Most of these flags are owned by the toolchain rather than by individual ports. At a minimum this needs a big comment describing what it's doing.

It also needs to at most add flags rather than clobbering them.

File: ports/ginkgo/portfile.cmake [hunk 3]
```diff
@@ -33,20 +33,14 @@ vcpkg_cmake_configure(
...
-        -DGinkgo_NAME=ginkgo
         ${FEATURE_OPTIONS}
+        ${ADDITIONAL_FLAGS}
 )
 
 vcpkg_cmake_install()
 vcpkg_cmake_config_fixup(CONFIG_PATH lib/cmake/Ginkgo)
 vcpkg_fixup_pkgconfig()
 
-vcpkg_replace_string("${CURRENT_PACKAGES_DIR}/share/ginkgo/GinkgoConfig.cmake" [[string(REPLACE "lib/cmake/Ginkgo" "" GINKGO_INSTALL_PREFIX "${GINKGO_CONFIG_FILE_PATH}")]] "")
```
Comment (BillyONeal) [inline ports/ginkgo/portfile.cmake:44]: Why did you delete these?

File: versions/g-/ginkgo.json [hunk 4]
```diff
@@ -1,5 +1,15 @@
...
+    {
+      "git-tree": "bf79f307113acd68629b6b715f774f3df07d8121",
+      "version-semver": "1.6.0",
+      "port-version": 0
+    },
+    {
+      "git-tree": "c28b801a453a01236309c6024dfb7cb111776444",
+      "version-semver": "1.5.0",
+      "port-version": 0
+    },
```
Comment (BillyONeal) [inline versions/g-/ginkgo.json:12]:
```suggestion
```

**Review by BillyONeal — DISMISSED** [review]
> x64_uwp fails because of SDL turning deprecation warnings into errors https://stackoverflow.com/questions/55007377/deprecated-results-in-build-failure-in-vs2017-error-c4996

You should patch the build or add /wd4996 as necessary to turn those warnings off. This change is probably fallout from `WINDOWS_USE_MSBUILD`.

File: ports/ginkgo/portfile.cmake (hunk 3 above)
Comment (upsj) [reply ports/ginkgo/portfile.cmake:44]: This should be fixed upstream, since we made the GinkgoConfig file relocatable with https://github.com/ginkgo-project/ginkgo/pull/1325

**Comment by upsj** [comment]
@BillyONeal It makes more sense to review https://github.com/microsoft/vcpkg/pull/27792 first, which is where most of the surprising changes originate.

File: ports/ginkgo/portfile.cmake (hunk 1 above)
Comment (upsj) [reply ports/ginkgo/portfile.cmake:24]: see https://github.com/microsoft/vcpkg/pull/27792#issuecomment-1628657995, this is related to our use of `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS` which we will not be able to fix without breaking interface (i.e. the next major release)

File: ports/ginkgo/portfile.cmake (hunk 2 above)
Comment (upsj) [reply ports/ginkgo/portfile.cmake]: I pulled the changed flags into a patch

File: ports/ginkgo/portfile.cmake (hunk 1 above)
Comment (BillyONeal) [reply ports/ginkgo/portfile.cmake:20]: Hmmm I'm not sure I understand. 1.4.0 is building with ninja just fine, and this is *introducing* a breaking change. Is it something specifically that happened in 1.5.0 or 1.6.0 that changed the situation?

File: ports/ginkgo/portfile.cmake (hunk 3 above)
Comment (BillyONeal) [reply ports/ginkgo/portfile.cmake:44]: Makes sense, thank you!

**Comment by BillyONeal** [comment]
Marked `vcpkg-team-review` because I want a second opinion on the `WINDOWS_USE_MSBUILD` thing.

File: ports/ginkgo/portfile.cmake (hunk 1 above)
Comment (upsj) [reply ports/ginkgo/portfile.cmake:20]: Not sure how familiar you are with `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS`, so I'll try to give some context: With symbols not being exported by default in MSVC, CMake added a workaround that scans the generated object files and stores all symbols they contain in an `exports.def` file to pass on to the linker (?) to export. This allows MSVC builds to work exactly like on Linux, where symbols are visible/exported by default. The downside of this approach is that it forces the use of 16 bit ordinals for referring to the symbols instead of actual named exports (as far as I understand). With 1.5.0 and 1.6.0 we added a significant amount of features and thus symbols, so dealing with MSVC was always a fight against breaking the 65536 exported symbol count.

What is going on with Ninja here I cannot entirely answer, but from observing the build process, it seems to me like MSBuild is compiling multiple files in a single CL.exe invocation, which despite otherwise using the same flags, leads to less symbols being present in the object files compared to Ninja (IPO?), which launches a single CL.exe process for source file. Essentially it's a stopgap measure for debug builds until we manage to introduce `__dllexport/import` annotations in our code, but that requires some restructuring that involves an interface break.

Comment (BillyONeal) [reply ports/ginkgo/portfile.cmake:20]: @JavierMatosD @ras0219-msft @data-queue @vicroms and I looked at this today

> fight against breaking the 65536 exported symbol count

Could we fix that by just saying this port only produces static libraries on Windows, or is there some reason it must support DLLs? Something like this might work:

```cmake
if(VCPKG_TARGET_IS_WINDOWS)
    vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
endif()
```

Note that vcpkg's usual policy for libraries that need `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS` is to say that those libraries functionally don't support being DLLs. See https://learn.microsoft.com/vcpkg/contributing/maintainer-guide#do-not-add-cmake_windows_export_all_symbols

Comment (upsj) [reply ports/ginkgo/portfile.cmake:20]: We still want to keep up the compatibility on our side, but maybe it's fine to ask people requiring DLLs to build the software manually. So yes, that is acceptable to me. Thanks for the input, I wasn't aware of the policy, or it wasn't there when I first added the port 😄

File: ports/ginkgo/portfile.cmake [hunk 5]
```diff
@@ -1,23 +1,20 @@
...
+if(VCPKG_TARGET_IS_WINDOWS)
+    vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
+endif()
 
 vcpkg_from_github(
     OUT_SOURCE_PATH SOURCE_PATH
     REPO ginkgo-project/ginkgo
-    REF v1.4.0
-    SHA512 9bfcb2c415c7a1a70cf8e49f20adf62d02cab83bb23b6fcecfeaeeb322b2d4e1ad8d8fa6582735073753f9a05eac8688b9bd1ff1d4203957c1a80702d117e807
+    REF v1.6.0
```
Comment (Adela0814) [inline ports/ginkgo/portfile.cmake:8]:
```suggestion
    REF "v${VERSION}"
```

**Review by Adela0814 — CHANGES_REQUESTED** [review]

**Comment by Adela0814** [comment]
I encountered the following error while testing all features:
Repro step:  `.\vcpkg.exe install ginkgo[*]:x64-windows`.
Error message:
```
CMake Error at E:/5/vcpkg/downloads/tools/cmake-3.27.1-windows/cmake-3.27.1-windows-i386/share/cmake-3.27/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find OpenMP_C: Found unsuitable version "2.0", but required is at
  least "3.0" (found -openmp)
Call Stack (most recent call first):
  E:/5/vcpkg/downloads/tools/cmake-3.27.1-windows/cmake-3.27.1-windows-i386/share/cmake-3.27/Modules/FindPackageHandleStandardArgs.cmake:598 (_FPHSA_FAILURE_MESSAGE)
  E:/5/vcpkg/downloads/tools/cmake-3.27.1-windows/cmake-3.27.1-windows-i386/share/cmake-3.27/Modules/FindOpenMP.cmake:577 (find_package_handle_standard_args)
  E:/5/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
  CMakeLists.txt:105 (find_package)
```
[config-x64-windows-out.log](https://github.com/microsoft/vcpkg/files/12891009/config-x64-windows-out.log)

**Review by Adela0814 — APPROVED** [review]

**Comment by BillyONeal** [comment]
> but it doesn't seem possible to express what features are available on which platforms except through the descriptions.

You should be able to add something like `"supports": "!windows | !mingw"` to the feature?

**Comment by BillyONeal** [comment]
Thanks for the update!
